src/fallbacks.jl

```diff
@@ -0,0 +1,121 @@
+    
+    Custom column types can override with an appropriate "scalar" element type that should dispatch to their column allocator.
+"""
+allocatecolumn(T, len) = Vector{T}(undef, len)
+
+@inline function allocatecolumns(::Schema{names, types}, len) where {names, types}
+    if @generated
+        vals = Tuple(:(allocatecolumn($(fieldtype(types, i)), len)) for i = 1:fieldcount(types))
+        return :(NamedTuple{names}(($(vals...),)))
+    else
+        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(types, i), len) for i = 1:fieldcount(types)))
+    end
+end
+
+haslength(x) = x === Base.HasLength() || x === Base.HasShape{1}()
+
+# add! will push! or setindex! a value depending on if the row-iterator HasLength or not
+@inline add!(val, col::Int, nm::Symbol, ::Union{Base.HasLength, Base.HasShape{1}}, nt, row) = setindex!(nt[col], val, row)
+@inline add!(val, col::Int, nm::Symbol, T, nt, row) = push!(nt[col], val)
+
+@inline function buildcolumns(schema, rowitr::T) where {T}
+    L = Base.IteratorSize(T)
+    len = haslength(L) ? length(rowitr) : 0
+    nt = allocatecolumns(schema, len)
+    for (i, row) in enumerate(rowitr)
+        eachcolumn(add!, schema, row, L, nt, i)
+    end
+    return nt
+end
+
+@inline add!(dest::AbstractVector, val, ::Union{Base.HasLength, Base.HasShape{1}}, row) = setindex!(dest, val, row)
+@inline add!(dest::AbstractVector, val, T, row) = push!(dest, val)
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, nm::Symbol, L, row, len, updated) where {T, S}
+    if S === T || val isa T
+        add!(dest, val, L, row)
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), max(len, length(dest)))
+        row > 1 && copyto!(new, 1, dest, 1, row - 1)
+        add!(new, val, L, row)
+        updated[] = merge(updated[], NamedTuple{(nm,)}((new,)))
+        return new
+    end
+end
+
+@inline function add_or_widen!(val, col, nm, L, columns, row, len, updated)
+    @inbounds add_or_widen!(columns[col], val, nm, L, row, len, updated)
+    return
+end
+
+# when Tables.schema(x) === nothing
+function buildcolumns(::Nothing, rowitr::T) where {T}
+    state = iterate(rowitr)
+    state === nothing && return NamedTuple()
+    row::eltype(rowitr), st = state
+    names = propertynames(row)
+    L = Base.IteratorSize(T)
+    len = haslength(L) ? length(rowitr) : 0
+    sch = Schema(names, nothing)
+    columns = NamedTuple{names}(Tuple(Union{}[] for _ = 1:length(names)))
+    return _buildcolumns(rowitr, row, st, sch, L, columns, 1, len, Ref{Any}(columns))
+end
+
+function _buildcolumns(rowitr, row, st, sch, L, columns, rownbr, len, updated)
+    while true
+        eachcolumn(add_or_widen!, sch, row, L, columns, rownbr, len, updated)
+        rownbr += 1
+        state = iterate(rowitr, st)
+        state === nothing && break
+        row, st = state
+        columns !== updated[] && return _buildcolumns(rowitr, row, st, sch, L, updated[], rownbr, len, updated)
```

Comment:
@piever, this is similar to the idea you proposed for making the iteration loop more type-stable. Here I'm passing in a `Ref{Any}(columns)` when we set/push each row value, and if a column needs to widen, it will update the ref w/ `merge(updated[], NamedTuple{(nm,)}((new_column,))`. A tricky thing that makes this work is the fact that all of the non-widened columns are identical objects between the `columns` and `updated[]` arguments; i.e. the `merge` call makes a new namedtuple that points to all the original columns except the one new widened column. And since we only update each column once per row, we can just check at the end of a row loop if the `updated[]` changed and re-dispatch on that.

This approach seems to perform really well on some larger datasets I tested, comparing the type-stable `columntable(f)` vs. `Tables.buildcolumns(nothing, f)`.

Reply:
Good trick! How about adding a comment? ;-)

Reply:
Yes! Definitely a "best of both worlds" implementation. I agree with @nalimilan that this needs comments in the code as it's a really smart strategy.